x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java

```diff
@@ -73,76 +74,108 @@ public SamlAuthnRequestValidator(SamlFactory samlFactory, SamlIdentityProvider i
+        }
+        logger.trace(new ParameterizedMessage("Parsed the following parameters from the query string: {}", parameters));
+        final String samlRequest = parameters.get("SAMLRequest");
+        if (null == samlRequest) {
+            throw new ElasticsearchSecurityException("Query string [{}] does not contain a SAMLRequest parameter", queryString);
         }
+        return new ParsedQueryString(
+            queryString,
+            samlRequest,
+            parameters.get("RelayState"),
+            parameters.get("SigAlg"),
+            parameters.get("Signature"));
+    }
+
+    private boolean validateAuthnRequest(AuthnRequest authnRequest, SamlServiceProvider sp, ParsedQueryString parsedQueryString,
+                                         ActionListener<SamlValidateAuthnRequestResponse> listener) {
+        // If the Service Provider should not sign requests, do not try to handle signatures even if they are added to the request
+        if (sp.shouldSignAuthnRequests()) {
+            if (Strings.hasText(parsedQueryString.signature)) {
+                if (Strings.hasText(parsedQueryString.sigAlg) == false) {
+                    logAndRespond(new ParameterizedMessage("Query string [{}] contains a Signature but SigAlg parameter is missing",
+                        parsedQueryString.queryString), listener);
+                    return true;
+                }
+                final X509Credential spSigningCredential = sp.getSpSigningCredential();
+                if (spSigningCredential == null) {
+                    logAndRespond(
+                        "Unable to validate signature of authentication request, " +
+                            "Service Provider hasn't registered signing credentials",
+                        listener);
+                    return true;
+                }
+                if (validateSignature(parsedQueryString, spSigningCredential) == false) {
+                    logAndRespond(
+                        new ParameterizedMessage("Unable to validate signature of authentication request [{}] using credentials [{}]",
+                            parsedQueryString.queryString,
+                            samlFactory.describeCredentials(Collections.singletonList(spSigningCredential))), listener);
+                    return true;
+                }
+            } else if (Strings.hasText(parsedQueryString.sigAlg)) {
+                logAndRespond(new ParameterizedMessage("Query string [{}] contains a SigAlg parameter but Signature is missing",
+                    parsedQueryString.queryString), listener);
+                return true;
+            } else {
+                logAndRespond("The Service Provider must sign authentication requests but no signature was found", listener);
+                return true;
+            }
+        }
+        validateAuthnRequest(authnRequest, sp);
```

Comment:
Again - I see the pattern here.. - I dislike the fact that we have multiple methods with the same name doing similar stuff. Arguments can be made that signature verification _is_ part of the validation of the AuthnRequest so I'm fine with how you moved that around but could we simple replace this with 

```suggestion
        checkDestination(authnRequest);
        checkAcs(authnRequest, sp);
```

? I see no reason keeping the second `validateAuthnRequest` now